### PR TITLE
Add to O.Paths the ability to exclude custom types and to get all available paths

### PR DESCRIPTION
**Pull request by regevbr**
## 🎁 Pull Request

* [x] Used a clear / meaningful title for this pull request
* [x] Tested the changes in your own code (on your projects)
* [x] Added / Edited tests to reflect changes (`tests` folder)
* [x] Have read the **Contributing** part of the **Readme**
* [x] Passed `npm test`

#### Fixes
This adds 2 features to `O.Paths`:
1. Add the ability to "break early" on custom types to avoid stack errors due to circular references in external types (such as `Moment`
1. Add the ability to actually get all the paths in an interface, and not the currently compressed version.
1. Add the ability to get the compact paths, but in a `required` mode. This will help avoid calling `List.Required` which utilizes a lot of computing power for TS :-(
1. Added an option to exclude some key types. This can help reduce the keys used to string only, which makes the output usable by the `String.Join` type

Please see the added tests for examples

#### Why have you made changes?
Having external types in my interface that contain circular dependencies (such as `Moment`) causes TS to break.
In addition, I wanted to create a string type that contains all the available paths in a type.

#### What changes have you made?
* Changed `O.Paths` to accept 3 more optional parameters

#### What tests have you updated?
* tested this in `tests/Object.ts`

#### Is there any breaking changes?
* [ ] Yes, I changed the public API & documented it
* [ ] Yes, I changed existing tests
* [ ] No,  I added to the public API & documented it
* [x] No,  I added to the existing tests
* [ ] I don't know

#### Anything else worth mentioning?
@millsp your CR and thoughts will be much appreciated :-)

### CODE REVIEW

File: tests/Object.ts [hunk 1]
```diff
@@ -1057,10 +1057,36 @@ type O_PATHS = {
...
+            h: {r: number}[],
+        };
+    };
+};
+
 checks([
+    // Default compact mode
     check<O.Paths<{'prop': {a: 1}[]}>, T.NonNullable<['prop'?, number?, 'a'?]>, Test.Pass>(),
     check<O.Paths<O_PATHS>, T.NonNullable<['a'?, 'a'?] | ['b'?, 'a'?, 'a'?] | ['b'?, 'b'?] | ['c'?]>, Test.Pass>(),
-    check<O.Paths<O[]>, O.Paths<O[]>, Test.Pass>(),
```
Comment (regevbr) [inline tests/Object.ts:1063]: I removed this as it fails after my changes. To make it work you need to change the magic number 12 to 8... Since this happens due to circular ref I don't think we should support it. @millsp thoughts?

File: sources/Object/Paths.ts [hunk 2]
```diff
@@ -1,41 +1,83 @@
...
+type PathMode = 'compact' | 'all' | 'required'
+
 /**
  * @hidden
  */
 type UnionOf<A> =
     A extends List
     ? A[number]
-    : A[keyof A]
+    : Exclude<A[keyof A], undefined>
```
Comment (regevbr) [inline sources/Object/Paths.ts:21]: This fixes an edge case (I added tests for it) that causes optional fields (ie `{a?:number}`) to be converted to paths that contain `List<Key>` which in turn case cause `String.Join` to fail and fallback to `string`

**Comment by regevbr** [comment]
@millsp did you have a chance to review this PR?

**Comment by millsp** [comment]
Sorry for the delays, will check soon. This looks like it could fix a number of issues. Thank you for finding this!

**Comment by regevbr** [comment]
@millsp hows it going? Did you manage to check out the PR?

**Comment by regevbr** [comment]
@millsp hows it going? Did you manage to check out the PR?

**Comment by regevbr** [comment]
@millsp hows it going? Did you manage to check out the PR?

**Comment by regevbr** [comment]
@millsp can you please CR?

**Comment by regevbr** [comment]
@millsp can you please CR?
